rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -1816,6 +1816,61 @@ static rmw_ret_t rmw_take_int(
   return RMW_RET_OK;
 }
 
+static rmw_ret_t rmw_take_seq(
+  const rmw_subscription_t * subscription,
+  size_t count,
+  rmw_message_sequence_t * message_sequence,
+  rmw_message_info_sequence_t * message_info_sequence,
+  size_t * taken)
+{
+  RET_NULL(taken);
+  RET_NULL(message_sequence);
+  RET_NULL(message_info_sequence);
+  RET_WRONG_IMPLID(subscription);
+
+  if (count > message_sequence->capacity) {
+    RMW_SET_ERROR_MSG("Insuffient capacity in message_sequence");
+    return RMW_RET_ERROR;
+  }
+
+  if (count > message_info_sequence->capacity) {
+    RMW_SET_ERROR_MSG("Insuffient capacity in message_info_sequence");
+    return RMW_RET_ERROR;
+  }
+
+  CddsSubscription * sub = static_cast<CddsSubscription *>(subscription->data);
+  RET_NULL(sub);
+
+  std::vector<dds_sample_info_t> infos(count);
+  auto ret = dds_take(sub->enth, message_sequence->data, infos.data(), count, count);
+
+  if (ret <= 0) {
+    return RMW_RET_ERROR;
+  }
+
+  for (int ii = 0; ii < ret; ++ii) {
+    const dds_sample_info_t & info = infos[ii];
+    rmw_message_info_t * message_info = &message_info_sequence->data[ii];
+
+    if (info.valid_data) {
+      if (message_info) {
+        message_info->publisher_gid.implementation_identifier = eclipse_cyclonedds_identifier;
+        memset(message_info->publisher_gid.data, 0, sizeof(message_info->publisher_gid.data));
+        assert(sizeof(info.publication_handle) <= sizeof(message_info->publisher_gid.data));
+        memcpy(
+          message_info->publisher_gid.data, &info.publication_handle,
+          sizeof(info.publication_handle));
+      }
```

Comment:
What's the else case here? Why would that be null, and if it is, is that ok?

Reply:
message_info can't be null, I dare say. What I am more concerned about is what happens when `!valid_data`: that's a case the ROS2 applications don't expect. In Cyclone's case, the sample will be all-zero (except for the key fields, but there are none of those in ROS2).

You only get those in some edge cases: when the last publisher disappears, when a publisher loses its liveliness, stuff like that. I think those need to be hidden from the caller, just like `rmw_take` does.

Reply:
@eboasson Since we aren't currently controlling any allocation here, would it be sufficient to rearrange the sequences to keep all of the valid samples/infos at the beginning of the sequence, and indicate with the `size` field how many are valid?

Reply:
I went ahead and implemented it that way in ef76435

Reply:
@mjcarroll my apologies for not responding, but yes, that works as long as the upper layers don't mind the pointers getting reordered. I suppose those pointers ultimately come from the application and I don't know what the expectation of applications will be in this.

Reply:
In general I would expect the upper layers to do the allocation outside of the message_sequence structure, and only populate the pointers for the purpose of retrieving from the middleware.  In that case, I can't imagine that reordering the pointers should cause substantial issues, as long as it is documented.